Code/Framework/AzCore/AzCore/Jobs/Internal/JobTypeEraser.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Jobs/Internal/JobTypeEraser.h>
+
+namespace AZ::Internal
+{
+    TypeErasedJob::TypeErasedJob(TypeErasedJob&& other) noexcept
+    {
+        if (!other.m_relocator || other.m_lambda != other.m_buffer)
+        {
+            // The type-erased lambda is trivially relocatable OR, the lambda is heap allocated
+            memcpy(this, &other, sizeof(TypeErasedJob));
+
+            if (other.m_lambda == other.m_buffer)
+            {
+                m_lambda = m_buffer;
+            }
+
+            // Prevent deletion in the event the lambda had spilled to the heap
+            other.m_lambda = nullptr;
+            return;
+        }
+
+        // At this point, we know the lambda was inlined
+        m_lambda = m_buffer;
+
+        m_invoker = other.m_invoker;
+        m_relocator = other.m_relocator;
+        m_destroyer = other.m_destroyer;
+
+        // We now own the lambda, so clear the moved-from job's destroyer
+        other.m_destroyer = nullptr;
+        other.m_invoker = nullptr;
+
+        m_relocator(m_buffer, other.m_buffer);
+    }
+
+    TypeErasedJob& TypeErasedJob::operator=(TypeErasedJob&& other) noexcept
+    {
+        if (this == &other)
+        {
+            return *this;
+        }
+
+        this->~TypeErasedJob();
+
+        new (this) TypeErasedJob{ AZStd::move(other) };
+
+        return *this;
+    }
+
+    TypeErasedJob::~TypeErasedJob()
+    {
+        if (m_lambda)
+        {
+            if (m_destroyer)
+            {
+                // The presence of m_destroyer indicates that the lambda is not trivially destructible
+                m_destroyer(m_lambda);
+            }
+
+            if (m_lambda != m_buffer)
+            {
+                // We've spilled the lambda into the heap, free its memory
```

Comment:
I could certainly put in a compile time switch and have it off initially to see if we actually need it

Reply:
I feel like this can be encapsulated behind a unique_ptr with a custom deleter.

```
struct JobDestroyer
{
    JobDestroyer() noexcept;
    JobDestroyer(JobDestroy_t destroyer) noexcept;
    void operator()(void* ptr);
    JobDestroy_t m_destroyer;
};

using JobDestroyerPtr = AZStd::unique_ptr<void*, JobDestroyer>;
```

Reply:
I'm less worried about the technical implementation and more about a lightweight job system running with bloated jobs. By not supporting it in the job system nobody is locked out of creating bloated jobs because it's always possible to create data on the heap first and pass the pointer to the job. That way it puts the performant lightweight jobs front-and-center and makes bloated jobs stick out and easier to find, while the current flexible approach would silently hide it. Those universal captures for lambda's can drag in a lot by accident.

Reply:
If we are going to put it behind a compile time switch, I think it should be on by default.
The `static_assert` can provide the user with guidance of how to handle the situation.

@AMZN-koppersr you should be clearer by what you mean about "bloat".
Normally bloat when dealing with the job system is about the amount of time a job runs, not about the storage size of the functor associated with the job. But in this case bloat is referring to the storage size of the job.

Reply:
@AMZN-koppersr I'm going to go with your original suggestion and remove the heap spill and just inline all the data. Gets rid of a pointer, gives us 8 bytes more to work with, gets rid of an indirection potentially and a number of branches.